File: python/selenium-proxy-session/scrape.py
#!/usr/bin/env python3
import json
from os import environ
from random import random
from selenium.webdriver import Remote, ChromeOptions as Options
from selenium.webdriver.chromium.remote_connection import ChromiumRemoteConnection as Connection
from selenium.webdriver.common.by import By

AUTH = environ.get('AUTH', default='USER:PASS')
TARGET_URL = environ.get('TARGET_URL', default='https://ifconfig.co/json')

def scrape(num, url, session_id):
    if AUTH == 'USER:PASS':
        raise Exception('Provide Scraping Browsers credentials in AUTH '
                        'environment variable or update the script.')
    print(f'Scrape {num}: Connecting to Browser...')
    server_addr = f'https://{AUTH}@brd.superproxy.io:9515'
    connection = Connection(server_addr, 'goog', 'chrome')
    driver = Remote(connection, options=Options())

    def cdp(cmd, params={}):
        return driver.execute('executeCdpCommand', {
            'cmd': cmd,
            'params': params,
        })['value']

    try:
        print(f'Scrape {num}: Connected! Attaching session {session_id}...')
        cdp('Proxy.useSession', {'sessionId': session_id})
        print(f'Scrape {num}: Navigating to {url}...')
        driver.get(url)
        print(f'Scrape {num}: Navigated! Scraping data...')
        body = driver.find_element(By.CSS_SELECTOR, 'body')
        data = body.text
        return json.loads(data)['ip']
    finally:
        print(f'Scrape {num}: Closing browser...')
        driver.quit()


def main(url=TARGET_URL):
    session_id = format(int(0xFFFFFFFF * random()), 'x')
    print(f'Using proxy session: {session_id}')
    ip1 = scrape(1, url, session_id)
    ip2 = scrape(2, url, session_id)
    print('IPs:', [ip1, ip2])
    print('Same IP (session persisted)?', ip1 == ip2)


if __name__ == '__main__':
    main()

File: nodejs/selenium-proxy-session/scrape.js
#!/usr/bin/env node
const { Builder, Browser, By } = require('selenium-webdriver');
const {
    AUTH = 'USER:PASS',
    TARGET_URL = 'https://ifconfig.co/json',
} = process.env;

async function scrape(num, url, sessionId) {
    if (AUTH == 'USER:PASS') {
        throw new Error(`Provide Scraping Browsers credentials in AUTH`
            + ` environment variable or update the script.`);
    }
    console.log(`Scrape ${num}: Connecting to Browser...`);
    const server = `https://${AUTH}@brd.superproxy.io:9515`;
    const driver = await new Builder()
        .forBrowser(Browser.CHROME)
        .usingServer(server)
        .build();
    try {
        console.log(`Scrape ${num}: Connected! Attaching session ${sessionId}...`);
        await driver.sendAndGetDevToolsCommand('Proxy.useSession', { sessionId });
        console.log(`Scrape ${num}: Navigating to ${url}...`);
        await driver.get(url);
        console.log(`Scrape ${num}: Navigated! Scraping data...`);
        const body = await driver.findElement(By.css('body'));
        const data = await body.getText();
        return JSON.parse(data).ip;
    } finally {
        console.log(`Scrape ${num}: Closing browser...`);
        await driver.quit();
    }
}

async function main(url = TARGET_URL) {
    const sessionId = Math.floor(0xFFFFFFFF * Math.random()).toString(16);
    console.log(`Using proxy session: ${sessionId}`);
    const ip1 = await scrape(1, url, sessionId);
    const ip2 = await scrape(2, url, sessionId);
    console.log(`IPs:`, [ip1, ip2]);
    console.log(`Same IP (session persisted)?`, ip1 === ip2);
}

function getErrorDetails(error) {
    if (error.target?._req?.res) {
        const {
            statusCode,
            statusMessage,
        } = error.target._req.res;
        return `Unexpected Server Status ${statusCode}: ${statusMessage}`;
    }
}

if (require.main == module) {
    main().catch(error => {
        console.error(getErrorDetails(error)
            || error.stack
            || error.message
            || error);
        process.exit(1);
    });
}
